src/Oscoin/Storage/BlockTree/SQLite.hs

```diff
@@ -0,0 +1,235 @@
+import           Oscoin.Consensus.Types (Validate)
+import           Oscoin.Crypto.Blockchain (Blockchain(..))
+import           Oscoin.Crypto.Blockchain.Block
+                 (Block(..), Depth, ScoreFn, Sealed, isGenesisBlock)
+import qualified Oscoin.Crypto.Blockchain.Block as Block
+import qualified Oscoin.Crypto.Hash as Crypto
+import           Oscoin.Telemetry.Events (NotableEvent(..))
+import           Oscoin.Time.Chrono as Chrono
+
+import           Oscoin.Storage.Block.SQLite.Internal (Handle)
+import qualified Oscoin.Storage.Block.SQLite.Internal as SQLite
+import qualified Oscoin.Storage.BlockTree.Internal as Abstract
+
+import           Oscoin.Storage.Block.Orphanage (Orphanage)
+import qualified Oscoin.Storage.Block.Orphanage as O
+
+import           Database.SQLite.Simple.FromField (FromField)
+import           Database.SQLite.Simple.ToField (ToField)
+
+import           Codec.Serialise (Serialise)
+import qualified Data.List.NonEmpty as NonEmpty
+import           Formatting
+
+-- | Bracket-style initialisation of the SQLite BlockTree
+-- FIXME(adn): This internally calls 'initialize', which is obviously not
+-- ideal when we have state on disk and we are re-opening a database rather
+-- than starting from scratch. In practice, /for now/, we should be OK as
+-- 'initialize' should be idempotent, but nevertheless we need a better
+-- migration strategy/initialisation here.
+withBlockTree
+    :: forall c tx s b.
+       ( ToField s
+       , FromField s
+       , FromField (Crypto.Hash c)
+       , Serialise s
+       , SQLite.StorableTx c tx
+       , Buildable (Crypto.Hash c)
+       , Eq s
+       )
+    => Consensus.Config
+    -> Validate c tx s
+    -- ^ A function to validate a block (or a chain of blocks)
+    -> ScoreFn c tx (Sealed c s)
+    -- ^ A function to score a block.
+    -> String
+    -- ^ The path where the DB will live on disk
+    -> Block c tx (Sealed c s)
+    -- ^ The genesis block (used to initialise the store)
+    -> (Abstract.BlockTree c tx s IO -> IO b)
+    -- ^ Action which uses the 'BlockTree'.
+    -> IO b
+withBlockTree config validateFull scoreBlock path genesisBlock action =
+    bracket acquire release use
+  where
+      acquire :: IO (MVar (Orphanage c tx s), Handle c tx s)
+      acquire = (,) <$> newMVar (O.emptyOrphanage scoreBlock)
+                    <*> (SQLite.initialize genesisBlock =<< SQLite.open path)
+
+      release :: (MVar (Orphanage c tx s), Handle c tx s) -> IO ()
+      release (_, hdl) = SQLite.close hdl
+
+      use :: (MVar (Orphanage c tx s), Handle c tx s)
+          -> IO b
+      use (orphansVar, internalHandle) =
+          let newBlockTree =
+                ( Abstract.BlockTreeReader {
+                      Abstract.getGenesisBlock       = SQLite.getGenesisBlock internalHandle
+                    , Abstract.lookupBlock           =
+                        -- NOTE(adn): This won't lookup orphans (currently). Should we
```

Comment:
By orphan you mean blocks without a parent, or blocks that aren't on the main chain?

Reply:
I was specifically thinking about blocks not on the main chain. In other terms, I quite like the fact the previous `BlockStoreReader` was a sort of "oracle" for the main chain, but if we now lift this property it means our queries are ambiguous, i.e. we wouldn't know if the requested block is on the chain or just an orphan. This in principle can be solved by returning a richer type (i.e `LookupBlockResult` or similar) but I do wonder if we ever need to care about orphans for such queries, and what would happen otherwise. Some food for thoughts:

* The `Ledger` calls `lookupBlock` and this means we would be able to materialise a block even if the parent is an orphan -- is this acceptable behaviour or do we want the `Ledger` to care only about main blocks? Wouldn't be wasteful to do otherwise?

* `lookupBlock` and `lookupTx` are used by the gossip layer to reply to queries (cfr. `wrapLookup`) and while in principle it makes sense to lookup both for orphans & main blocks, it's also true that probably the gossip layer will use this function to exchange main chain blocks during sync, so do we need to also be able to return back orphans (cc @kim ) ?

Reply:
I think if possible we should only give access to blocks on the main chain with `lookupBlock`. If later we want to give access to other blocks, we can add a `lookupAnyBlock` or something.

On another note, let's try to be more precise with terminology: an orphan is a block without a parent, while a *stale* block is a block not on the main chain.

Reply:
Gossip deals with _seen messages_, so it should be _all_ blocks seen so far. As mentioned before, this should be handled by a bloom filter maintained by the gossip layer, not read-through to storage.

Reply:
Right, ok, so it looks like the morally correct thing to do is to modify `lookupBlock` and `lookupTx` to deal exclusively with blocks on the main chain, and open an issue to extend the gossip layer to do the bookeeping on its side. Sounds like a plan?

> On another note, let's try to be more precise with terminology: an orphan is a block without a parent, while a stale block is a block not on the main chain.

Roger -- it also implies a subtle property: a stale block is a block which has "lost" against the main chain and that has been discarded, otherwise it would have been named simply "orphan". To say it differently, an orphan block is a block without a parent which at some point _might_ be considered for inclusion into the main chain. If that fails, it transitions from being an orphan to being stale.

Reply:
What's the significance of distinguishing stale from orphan blocks? `lookupBlock` will see stale blocks until the rollback transaction in the event of switching to a new chain commits.

Reply:
> a stale block is a block which has "lost" against the main chain and that has been discarded, otherwise it would have been named simply "orphan".

I don’t think it is correct to discard stale blocks: Assume we see two blocks _A_ and _B_ with a common ancestor _O_. Now _B_ scores higher than _A_ so we discard _A_. But the miner of the next block never saw _B_ so the next block _C_ is a parent of _A_ and not of _B_. Now the best chain is _O<-A<-C_. Thus _A_ needs to be re-integrated into the best chain and may not be discarded.

Reply:
> I don’t think it is correct to discard stale blocks: Assume we see two blocks A and B with a common ancestor O. Now B scores higher than A so we discard A. But the miner of the next block never saw B so the next block C is a parent of A and not of B. Now the best chain is O<-A<-C. Thus A needs to be re-integrated into the best chain and may not be discarded.

I guess this is part of a bigger discussion (which we might need to spin up separately) but I think the scenario you are describing shouldn't happen in practice -- at least according to "Mastering Bitcoin" (Chapter 10: Mining and consensus, page 244) it reads:

>Any miners working on extending the chain star-upside-down-triangle (i.e. O<-A) will now stop that work because their candidate block is an "orphan", as its parent "upside-down-triangle" (i.e. A) is no longer on the longest chain.

However I agree that this stuff is trick and easy to get wrong: another use case for a spec to state down clearly our consensus rules and how we plan to address forks and under which rules orphan/stale blocks are governed.

Reply:
> Any miners working on extending the chain star-upside-down-triangle (i.e. O<-A) will now stop that work because their candidate block is an "orphan", as its parent "upside-down-triangle" (i.e. A) is no longer on the longest chain.

This works if the miner receives _B_ before it mines a block. But we have to consider the case that it does not receive a block until after it has mined _C_.

Reply:
The cited example makes it quite clear. “upside-down triangle” corresponds to _B_ and “triangle” corresponds to _A_. The book tells us that “nodes will set the chain star-triangle-rhombus as the main chain and change the star-upside-down-triangle chain to a secondary chain.” This means that triangle cannot be discarded.

Reply:
> What's the significance of distinguishing stale from orphan blocks? `lookupBlock` will see stale blocks until the rollback transaction in the event of switching to a new chain commits.

From an external API perspective, none, but internally it makes sense as they have different properties and might be stored and retrieved differently, eg. an orphan block cannot have its state materialized and thus cannot be fully validated, whereas a stale block can.